src/LSST/system/ComConnection.cpp

```diff
@@ -0,0 +1,137 @@
+#include <ctime>
+#include <functional>
+#include <iostream>
+#include <stdexcept>
+
+// Third party headers
+
+// Project headers
+#include "system/ComServer.h"
+#include "system/Config.h"
+#include "system/Log.h"
+
+using namespace std;
+using namespace std::placeholders;
+
+namespace {
+
+bool isErrorCode(boost::system::error_code const& ec, string const& note) {
+    using Log = LSST::m2cellcpp::system::Log;
+    if (ec.value() != 0) {
+        if (ec == boost::asio::error::eof) {
+            Log::log(Log::INFO, note + "  ** closed **");
+        } else {
+            Log::log(Log::ERROR, note + "  ** failed: " + ec.message() + " **");
+        }
+        return true;
+    }
+    return false;
+}
+
+}  // namespace
+
+namespace LSST {
+namespace m2cellcpp {
+namespace system {
+
+ComConnection::Ptr ComConnection::create(IoContextPtr const& ioContext, uint64_t connId,
+                                         shared_ptr<ComServer> const& server) {
+    return ComConnection::Ptr(new ComConnection(ioContext, connId, server));
+}
+
+ComConnection::ComConnection(IoContextPtr const& ioContext, uint64_t connId, shared_ptr<ComServer> const& server) 
+  : _socket(*ioContext), _ioContext(ioContext), _connId(connId), _server(server) {
+}
+
+ComConnection::~ComConnection() {
+    shutdown();
+}
+
+void ComConnection::beginProtocol() { _receiveCommand(); }
+
+void ComConnection::_receiveCommand() {
+    Log::log(Log::DEBUG, "ComConnection::_receiveCommand");
+    if (_shutdown) {
+        return;
+    }
+    boost::asio::async_read_until(_socket, _streamBuf, getDelimiter(),
+                                  bind(&ComConnection::_readCommand, shared_from_this(), _1, _2));
+}
+
+void ComConnection::_readCommand(boost::system::error_code const& ec, size_t xfer) {
+    assert(_streamBuf.size() >= xfer);
+
+    if (::isErrorCode(ec, __func__)) return;
+    size_t msgSz = xfer - getDelimiter().size();
+    std::string command(buffers_begin(_streamBuf.data()), buffers_begin(_streamBuf.data()) + msgSz);
+    _streamBuf.consume(xfer);
+
+    Log::log(Log::INFO, "received command: " + command + " streamBuf size=" + to_string(command.size()));
+
+    // TODO: put command on a queue to be hanlded
+
+    // TODO: instead of echo, send response that the command was received
+    //      (will there be another response when the command is completed?)
+    _sendResponse(command);
+}
+
+void ComConnection::_sendResponse(string const& command) {
+    Log::log(Log::DEBUG, "ComConnection::_sendResponse command:" + command);
+    string cmd = command + ComConnection::getDelimiter();
+    boost::asio::async_write(_socket, boost::asio::buffer(cmd, cmd.size()),
+                             bind(&ComConnection::_responseSent, shared_from_this(), _1, _2));
+}
+
+void ComConnection::_responseSent(boost::system::error_code const& ec, size_t xfer) {
+    Log::log(Log::DEBUG, "ComConnection::_responseSent xfer=" + to_string(xfer));
+    if (::isErrorCode(ec, __func__)) return;
```

Comment:
I prefer something like:

```
if (::isErrorCode(ec, __func__)) {
    return;
}
```

But this is more like personal taste. However, the above method helps people who are novice of C++ like me.

Reply:
I prefer that as well. This just slipped by.